File: web/src/data/NoteStatsRepository.ts
import {
  collection,
  doc,
  onSnapshot,
  setDoc,
  serverTimestamp,
  increment,
  type Firestore,
  type Unsubscribe,
} from 'firebase/firestore'
import type { Auth } from 'firebase/auth'
import { noteStatsFromFirestore, type NoteStats } from './NoteStats'
import { firestoreUsage } from './FirestoreUsage'

function sameStats(a: Map<string, NoteStats>, b: Map<string, NoteStats>): boolean {
  if (a.size !== b.size) return false
  for (const [id, av] of a) {
    const bv = b.get(id)
    if (!bv) return false
    if (av.viewCount !== bv.viewCount) return false
    const aTs = av.lastAccessedAt
    const bTs = bv.lastAccessedAt
    if ((aTs == null) !== (bTs == null)) return false
    if (aTs != null && bTs != null && !aTs.isEqual(bTs)) return false
    const aDays = Object.keys(av.viewedDays)
    const bDays = Object.keys(bv.viewedDays)
    if (aDays.length !== bDays.length) return false
    for (const day of aDays) if (!bv.viewedDays[day]) return false
  }
  return true
}

/**
 * Per-user view tracking. Lives apart from the note doc so writes don't echo through
 * the notes collection listener and trigger the directive cache invalidation path.
 *
 * Reads are served from a listener-backed cache (lazy attach, full-collection
 * map, in-memory lookup); writes go directly to Firestore. See
 * `docs/firestore-efficiency.md` Principle 1. Mirrors the [RecentTabsRepository]
 * pattern. Use the module singleton [noteStatsRepo] — two listeners on the
 * same `noteStats` collection would double the listener-read count.
 *
 * Lifecycle: db/auth are bound by FirestoreLifecycle attach/detach so the
 * SDK can be terminated on idle. Subscribers and pending load promises
 * survive a detach so wake-ups re-load transparently.
 */
export class NoteStatsRepository {
  private db: Firestore | null = null
  private auth: Auth | null = null
  private cachedStats: Map<string, NoteStats> = new Map()
  private unsubscribe: Unsubscribe | null = null
  private loadPromise: Promise<void> | null = null
  private loadResolve: (() => void) | null = null
  private subscribers = new Set<() => void>()

  attach(db: Firestore, auth: Auth): void {
    this.db = db
    this.auth = auth
    // The listener is lazy (attached on first loadAllNoteStats); nothing to
    // do here. If a load was pending across a detach, it'll resume next call.
  }

  detach(): void {
    this.unsubscribe?.()
    this.unsubscribe = null
    this.loadPromise = null
    this.loadResolve = null
    this.cachedStats = new Map()
    this.db = null
    this.auth = null
    // Subscribers are kept — same React subscribers will receive updates when
    // the listener is reattached.
  }

  private requireDb(): Firestore {
    if (!this.db) throw new Error('NoteStatsRepository not attached (Firestore lifecycle stopped)')
    return this.db
  }

  private requireUserId(): string {
    if (!this.auth) throw new Error('NoteStatsRepository not attached')
    const uid = this.auth.currentUser?.uid
    if (!uid) throw new Error('User not signed in')
    return uid
  }

  private statsCollection(userId: string) {
    return collection(this.requireDb(), 'users', userId, 'noteStats')
  }

  private statsRef(userId: string, noteId: string) {
    return doc(this.statsCollection(userId), noteId)
  }

  private ensureListenerAttached(): Promise<void> {
    if (this.loadPromise) return this.loadPromise
    const userId = this.requireUserId()

    this.loadPromise = new Promise<void>((resolve) => {
      this.loadResolve = resolve
    })

    this.unsubscribe = onSnapshot(
      this.statsCollection(userId),
      (snapshot) => {
        const isFirstSnapshot = this.loadResolve !== null
        const isLocalEcho = !isFirstSnapshot && snapshot.metadata.hasPendingWrites
        const fromCache = snapshot.metadata.fromCache
        const type = isFirstSnapshot
          ? fromCache
            ? 'LISTENER_INITIAL_CACHED'
            : 'LISTENER_INITIAL_FRESH'
          : isLocalEcho
            ? 'LISTENER_LOCAL_ECHO'
            : fromCache
              ? 'LISTENER_UPDATE_CACHED'
              : 'LISTENER_UPDATE_FRESH'
        const docCount = isFirstSnapshot ? snapshot.docs.length : snapshot.docChanges().length
        firestoreUsage.recordRead('NoteStatsRepo.listener', type, docCount)

        const next = new Map<string, NoteStats>()
        snapshot.docs.forEach((d) => {
          next.set(d.id, noteStatsFromFirestore(d.data()))
        })
        const changed = !sameStats(this.cachedStats, next)
        if (changed) this.cachedStats = next

        if (isFirstSnapshot) {
          this.loadResolve?.()
          this.loadResolve = null
        }
        if (changed) {
          for (const sub of this.subscribers) sub()
        }
      },
      (error) => {
        console.error('[NoteStatsRepo] listener failed', error)
        if (this.loadResolve) {
          this.loadResolve()
          this.loadResolve = null
        }
      },
    )

    return this.loadPromise
  }

  /** Subscribe to cache updates. Returns an unsubscribe function. */
  subscribe(listener: () => void): () => void {
    this.subscribers.add(listener)
    return () => this.subscribers.delete(listener)
  }

  /** Embedded `[view ...]` directive renders MUST NOT call this — only top-level note opens. */
  async recordView(noteId: string): Promise<void> {
    const userId = this.requireUserId()
    const today = todayLocalDate()
    await setDoc(
      this.statsRef(userId, noteId),
      {
        lastAccessedAt: serverTimestamp(),
        viewCount: increment(1),
        viewedDays: { [today]: true },
      },
      { merge: true },
    )
    firestoreUsage.recordWrite('recordView', 'SET')
  }

  async loadAllNoteStats(): Promise<Map<string, NoteStats>> {
    await this.ensureListenerAttached()
    return this.cachedStats
  }
}

function todayLocalDate(): string {
  const d = new Date()
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${y}-${m}-${day}`
}

/**
 * Process-wide singleton. Two listeners on the same `noteStats` collection
 * would double the listener-read count and the per-snapshot work; share one.
 *
 * The instance is bound to a live Firestore via `attach()` by the bootstrap
 * module on every lifecycle.start(). Tests construct their own instance and
 * call attach() with a mocked Firestore directly.
 */
export const noteStatsRepo = new NoteStatsRepository()

File: app/src/main/java/org/alkaline/taskbrain/data/NoteStatsRepository.kt
package org.alkaline.taskbrain.data

import android.util.Log
import androidx.annotation.VisibleForTesting
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.SetOptions
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

/**
 * Per-user view tracking. Lives apart from the note doc so writes don't echo through
 * the notes collection listener and trigger the directive cache invalidation path.
 *
 * Listener-backed cache, singleton-on-companion. See [AlarmRepository] for
 * the pattern (lazy attach, first-instance binding, [clear] required in
 * test `@Before`).
 */
class NoteStatsRepository(
    private val db: FirebaseFirestore = FirebaseFirestore.getInstance(),
    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
) {
    private fun requireUserId(): String =
        auth.currentUser?.uid ?: throw IllegalStateException("User not signed in")

    private fun statsCollection(userId: String) =
        db.collection("users").document(userId).collection("noteStats")

    private fun statsRef(userId: String, noteId: String) =
        statsCollection(userId).document(noteId)

    private fun ensureListenerAttached(): CompletableDeferred<Unit> {
        synchronized(cacheLock) {
            if (listener != null) return loadDeferred!!
            val userId = auth.currentUser?.uid
                ?: throw IllegalStateException("User not signed in")
            val deferred = loadDeferred ?: CompletableDeferred<Unit>().also { loadDeferred = it }

            listener = statsCollection(userId)
                .addSnapshotListener { snapshot, error ->
                    if (error != null) {
                        Log.e(
                            TAG,
                            "[noteStats listener failed] userId=$userId\n${error.stackTraceToString()}",
                            error,
                        )
                        NoteStore.raiseWarning(
                            "Note view-stats sync failed. Sort orders that depend on view " +
                                "history (Recent, Decayed, Consistency) may be stale until you " +
                                "restart the app. Check logcat tag '$TAG'."
                        )
                        if (!deferred.isCompleted) deferred.complete(Unit)
                        return@addSnapshotListener
                    }
                    if (snapshot == null) return@addSnapshotListener
                    val firstAlreadyDelivered = deferred.isCompleted
                    val isLocalEcho = firstAlreadyDelivered && snapshot.metadata.hasPendingWrites()
                    val fromCache = snapshot.metadata.isFromCache
                    val type = when {
                        !firstAlreadyDelivered && fromCache -> FirestoreUsage.ReadType.LISTENER_INITIAL_CACHED
                        !firstAlreadyDelivered -> FirestoreUsage.ReadType.LISTENER_INITIAL_FRESH
                        isLocalEcho -> FirestoreUsage.ReadType.LISTENER_LOCAL_ECHO
                        fromCache -> FirestoreUsage.ReadType.LISTENER_UPDATE_CACHED
                        else -> FirestoreUsage.ReadType.LISTENER_UPDATE_FRESH
                    }
                    val docCount = if (firstAlreadyDelivered) snapshot.documentChanges.size else snapshot.size()
                    FirestoreUsage.recordRead("NoteStatsRepo.listener", type, docCount)
                    _cachedStats.value = snapshot.documents.associate { doc ->
                        doc.id to parseStats(doc.data ?: emptyMap())
                    }
                    if (!deferred.isCompleted) deferred.complete(Unit)
                }
            return deferred
        }
    }

    /** Embedded `[view ...]` directive renders MUST NOT call this — only top-level note opens. */
    suspend fun recordView(noteId: String): Result<Unit> = runCatching {
        withContext(Dispatchers.IO) {
            val userId = requireUserId()
            val today = todayLocalDate()
            val updates = mapOf(
                "lastAccessedAt" to FieldValue.serverTimestamp(),
                "viewCount" to FieldValue.increment(1),
                "viewedDays" to mapOf(today to true),
            )
            statsRef(userId, noteId).set(updates, SetOptions.merge()).await()
            FirestoreUsage.recordWrite("recordView", FirestoreUsage.WriteType.SET)
            Unit
        }
    }.onFailure { Log.e(TAG, "Error recording view", it) }

    suspend fun loadAllNoteStats(): Result<Map<String, NoteStats>> = runCatching {
        ensureListenerAttached().await()
        _cachedStats.value
    }.onFailure { Log.e(TAG, "Error loading note stats", it) }

    @Suppress("UNCHECKED_CAST")
    private fun parseStats(data: Map<String, Any?>): NoteStats = NoteStats(
        lastAccessedAt = data["lastAccessedAt"] as? Timestamp,
        viewCount = (data["viewCount"] as? Long ?: 0).toInt(),
        viewedDays = (data["viewedDays"] as? Map<String, Boolean>) ?: emptyMap(),
    )

    private fun todayLocalDate(): String =
        LocalDate.ofInstant(Instant.now(), ZoneId.systemDefault()).toString()

    companion object {
        private const val TAG = "NoteStatsRepository"

        private val cacheLock = Any()
        private val _cachedStats = MutableStateFlow<Map<String, NoteStats>>(emptyMap())
        private var listener: ListenerRegistration? = null
        private var loadDeferred: CompletableDeferred<Unit>? = null

        /**
         * Detach the listener and drop the cache. Call on sign-out so the
         * next signed-in user doesn't see the previous user's stats.
         */
        fun clear() {
            synchronized(cacheLock) {
                listener?.remove()
                listener = null
                loadDeferred = null
                _cachedStats.value = emptyMap()
            }
        }

        /** Test seam: populate the cache directly, bypassing the listener. */
        @VisibleForTesting
        internal fun injectCacheForTest(stats: Map<String, NoteStats>) {
            synchronized(cacheLock) {
                if (listener == null) listener = ListenerRegistration { /* test seam */ }
                val deferred = loadDeferred ?: CompletableDeferred<Unit>().also { loadDeferred = it }
                _cachedStats.value = stats
                if (!deferred.isCompleted) deferred.complete(Unit)
            }
        }
    }
}
